2016/exercises/final_project/Alba-Petar-Rebecca-Simon/PolytopeProject/Gale_Diagram.cpp

```diff
@@ -0,0 +1,322 @@
+                        }
+                        facetCounter++;
+                        
+                        igraph_add_edges(&graph, &edges, 0);
+                        
+                        //writeGraph(graph, ("counting"+to_string(counter)+".txt").c_str());
+                        
+                        igraph_vector_destroy(&edges);
+                    }
+                }
+            }
+        }
+    }
+    
+    igraph_to_undirected(&graph, IGRAPH_TO_UNDIRECTED_COLLAPSE, 0);
+    
+    
+    return graph;
+}
+
+bool Gale_Diagram::isIsomorphic(const Gale_Diagram &gale) const{
+    igraph_bool_t result = false;
+    igraph_t graph = makeVertexFacetStructure();
+    igraph_t otherGraph = gale.makeVertexFacetStructure();
+    igraph_isomorphic(&graph, &otherGraph, &result);
+    return result;
+}
+
+//For any three points, looks whether they are on a line.
+bool Gale_Diagram::isSimplicial() const{
+    int s = (int)points.size();
+    for(int i=0; i<s-3; i++){
+        for(int j=i+1; j<s-2; j++){
+            //If ij is a vertical line
+            if(points[i].get_coordinates()[0]==points[j].get_coordinates()[0]){
+                for(int k=j+1; k<s-1; k++){
+                    if(points[k].get_coordinates()[0]==points[i].get_coordinates()[0]) return false;
+                }
+                continue;
+            }
+            
+            //check if k is on line ij
+            Point slopeCoordinates = points[i].minus(points[j]);
+            mpq_class slope = slopeCoordinates.get_coordinates()[1]/slopeCoordinates.get_coordinates()[0];
+            slope.canonicalize();
+            
+            mpq_class aSlope = slope*points[i].get_coordinates()[0];
+            aSlope.canonicalize();
+            aSlope -= points[i].get_coordinates()[1];
+            aSlope.canonicalize();
+            
+            for(int k=j+1; k<s-1; k++){
+                mpq_class rightHandSide = slope*points[k].get_coordinates()[0];
+                rightHandSide.canonicalize();
+                
+                rightHandSide-=aSlope;
+                rightHandSide.canonicalize();
+                
+                if(points[k].get_coordinates()[1] == rightHandSide){
+                    return false;
+                }
+            }
+        }
+    }
+    return true;
+}
+
+
+
+
+Matrix Gale_Diagram::galeToPolytope() const{
+    //Make points homogene: positive points on z=1 and negative points on z=-1
+    vector<vector<mpq_class>> B;
+    for (int i = 0; i < 8; ++i){
+        vector<mpq_class> homogeneCoordinates = points[i].get_coordinates();
+        if(points[i].get_sign()) {
+            homogeneCoordinates.push_back(1);
+        }
+        else{
+            homogeneCoordinates.push_back(-1);
+        }
+        B.push_back(homogeneCoordinates);
+    }
+    
+    Matrix BT = Matrix(B).transpose();
+    vector<vector<mpq_class>> Btransposed=BT.getMatrix();
```

Comment:
Well, actually in this case I don't think it'll really matter, because you didn't prepare your implementation of `Matrix` to take advantage of `std::move`. This addition to the language is even more important than anonymous functions, and required knowledge. After reading up on that, I'm sure you'll find many ways to improve your code!